File: common/src/main/java/net/impactdev/gts/common/plugin/GTSPlugin.java
package net.impactdev.gts.common.plugin;

import com.google.gson.Gson;
import net.impactdev.gts.api.environment.Environment;
import net.impactdev.gts.common.config.ConfigProvider;
import net.impactdev.gts.common.messaging.InternalMessagingService;
import net.impactdev.gts.common.plugin.bootstrap.GTSBootstrap;
import net.impactdev.impactor.api.Impactor;
import net.impactdev.impactor.api.plugin.ImpactorPlugin;
import net.impactdev.gts.api.extensions.ExtensionManager;
import net.impactdev.gts.api.storage.GTSStorage;

import java.io.InputStream;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public interface GTSPlugin extends ImpactorPlugin {

	static GTSPlugin instance() {
		return Impactor.getInstance().getRegistry().get(GTSPlugin.class);
	}

	<T extends GTSPlugin> T as(Class<T> type);

	/**
	 * Gets the bootstrapper that created this plugin instance.
	 *
	 * @return The bootstrapper responsible for loading the plugin.
	 */
	GTSBootstrap bootstrap();

	/**
	 * A helpful utility for providing configuration links used by the plugin.
	 * Outside providing a path to the language configuration, the provider
	 * additionally provides a cached version of the main configuration without
	 * the need of processing through {@link #config()}, where the config is
	 * optionally present.
	 *
	 * @return A utility class meant to help provide cached instances of the
	 * configuration used by GTS.
	 */
	ConfigProvider configuration();

	/**
	 * Fetches information regarding the environment the plugin is presently running on. This
	 * should contain information regarding the server type and branding, as well as version information
	 * GTS will care about.
	 *
	 * @return The environment respective to the current platform
	 */
	Environment environment();

	/**
	 *
	 * @return
	 */
	Gson gson();

	/**
	 *
	 * @return
	 */
	GTSStorage storage();

	ExtensionManager extensionManager();

	InternalMessagingService messagingService();

	CompletableFuture<String> playerDisplayName(UUID id);

	default InputStream resource(String path) {
		return this.getClass().getClassLoader().getResourceAsStream(path);
	}
}

File: common/src/main/kotlin/net/impactdev/gts/common/plugin/GTSPlugin.kt
package net.impactdev.gts.common.plugin

import com.google.common.collect.ImmutableList
import com.google.gson.Gson
import net.impactdev.gts.api.environment.Environment
import net.impactdev.gts.api.extension.ExtensionManager
import net.impactdev.gts.api.storage.GTSStorage
import net.impactdev.gts.common.messaging.InternalMessagingService
import net.impactdev.gts.common.plugin.bootstrap.GTSBootstrap
import net.impactdev.impactor.api.Impactor
import net.impactdev.impactor.api.plugin.ImpactorPlugin
import net.impactdev.impactor.api.plugin.components.Configurable
import net.impactdev.impactor.api.plugin.components.Depending
import net.impactdev.impactor.api.plugin.components.Translatable
import net.impactdev.impactor.api.storage.StorageType
import java.io.InputStream
import java.util.*

interface GTSPlugin : ImpactorPlugin, Configurable, Depending, Translatable {
    fun <T : GTSPlugin?> `as`(type: Class<T>?): T

    /**
     *
     *
     * @return
     */
    val bootstrap: GTSBootstrap

    /**
     * Fetches information regarding the environment
     *
     * @return
     */
    val environment: Environment?

    /**
     *
     * @return
     */
    val gson: Gson

    /**
     *
     * @return
     */
    val storage: GTSStorage
    val extensionManager: ExtensionManager?
    val messagingService: InternalMessagingService?
    val multiServerCompatibleStorageOptions: ImmutableList<StorageType?>
    fun getPlayerDisplayName(id: UUID?): String
    fun getResourceStream(path: String?): InputStream {
        return this.javaClass.classLoader.getResourceAsStream(path)
    }

    companion object {
        @kotlin.jvm.JvmStatic
		val instance: GTSPlugin
            get() = Impactor.getInstance().registry.get(GTSPlugin::class.java)
    }
}